### ISSUE TITLE
SimpleChains docs example fails with Enzyme illegal type analysis

### ISSUE THREAD

**Issue by ChrisRackauckas-Claude** [JuliaDiffEq/DiffEqSensitivity.jl#1534]
During local docs testing for SciML/SciMLSensitivity.jl#1490, `docs/src/examples/neural_ode/simplechains.md` failed while Documenter was executing example blocks.

Observed locally on Julia 1.12.6 during the full docs command:

```sh
timeout 3600 env JULIA_DEPOT_PATH="$PWD/.julia_depot" /home/crackauc/.juliaup/bin/julia +1.12.6 --project -e 'println("--- Instantiating project"); using Pkg; Pkg.instantiate(); Pkg.activate("docs"); Pkg.instantiate(); push!(LOAD_PATH, @__DIR__); println("+++ Building documentation"); include("docs/make.jl")'
```

Documenter reported the failing example at:

```text
docs/src/examples/neural_ode/simplechains.md:70-87
```

The error was:

```text
IllegalTypeAnalysisException: Enzyme compilation failed due to illegal type analysis
```

The stack involved `SimpleChains.SimpleChain`, `VectorizationBase.align`, and the example optimizer using:

```julia
OPT.AutoEnzyme(mode = Enzyme.set_runtime_activity(Enzyme.Reverse))
```

This file was not changed by #1490, so it appears to be an existing docs/example failure rather than a regression from that PR. It should be isolated from the #1490 doctest cleanup.

**Comment by ChrisRackauckas-Claude** [comment]
The SimpleChains blocker now has a draft upstream fix: PumasAI/SimpleChains.jl#220.

The extension installs a package-level Enzyme reverse rule and avoids differentiating through SimpleChains' stack-backed implementation. A multi-layer VJP in both regular reverse mode and runtime-activity reverse mode exactly matched the existing ChainRules pullback. Once that PR is merged and released, the SimpleChains example can be moved to the Enzyme backend without relying on an Enzyme compiler workaround.

**Comment by ChrisRackauckas-Claude** [comment]
Verified the draft SimpleChains fix end-to-end against current SciMLSensitivity `master`: with PumasAI/SimpleChains.jl#220 developed into the docs environment (Julia 1.12.6, Enzyme 0.13.182), an isolated Documenter run of the actual `docs/src/examples/neural_ode/simplechains.md` page completed all 300 Enzyme training steps and rendered HTML successfully. The loss decreased from `400.6504f0` to `0.49608016f0`; the reported `IllegalTypeAnalysisException` did not recur. No SciMLSensitivity source change is needed beyond consuming a SimpleChains release containing the rule.

**Comment by ChrisRackauckas-Claude** [comment]
Re-verified with the newest Enzyme — **still broken, and now definitively blocked on SimpleChains, not Enzyme**.

**Environment**: Julia 1.12.6, Enzyme v0.13.186, SimpleChains v0.4.8, StaticArrays (latest registry).

The SciML-free reduction (from EnzymeAD/Enzyme.jl#3292) still fails identically:

```
loss = 1.373889
ERROR: IllegalTypeAnalysisException: Enzyme compilation failed due to illegal type analysis.
 Failure within method: MethodInstance for loss(::SimpleChains.SimpleChain{...}, ::SVector{2, Float32}, ::Vector{Float32})
 [2] align @ VectorizationBase/src/alignment.jl:13
 [3] with_stack_memory @ SimpleChains/src/memory.jl:25
```

Status of the dependency chain:
- EnzymeAD/Enzyme.jl#3292 was closed by @wsmoses as expected behavior — SimpleChains' `with_stack_memory` intentionally reinterprets stack memory at the "wrong" type, so Enzyme is right to refuse; the fix belongs in SimpleChains as a custom rule.
- PumasAI/SimpleChains.jl#220 (package-level EnzymeCore reverse rule) is that fix and was already verified end-to-end against the actual `simplechains.md` docs page (300 training steps, loss 400.65 → 0.496). It is still **open/unmerged**.

So this issue stays open until SimpleChains#220 merges and a release is tagged; then the docs example just needs the dep bump and can move to Enzyme with no SciMLSensitivity change.

**Comment by ChrisRackauckas-Claude** [comment]
**The underlying failure is fixed and verified**: SimpleChains v0.5.0 (released 2026-08-25) ships the EnzymeCore extension from PumasAI/SimpleChains.jl#220, which makes the chain the differentiation boundary and eliminates the `IllegalTypeAnalysisException`.

Verified against current SciMLSensitivity `master` with an isolated Documenter build of the actual `docs/src/examples/neural_ode/simplechains.md` (Julia 1.12.7, SimpleChains v0.5.0, Enzyme v0.13.199, `warnonly = false`): all 300 `AutoEnzyme` training steps completed, printed loss decreased `225.3562f0` → `0.35378605f0`, and HTML rendering succeeded.

What remains before the real docs build picks this up (it instantiates the committed `docs/Manifest.toml`, currently pinning SimpleChains v0.4.8):

1. Lux must admit SimpleChains 0.5 — every released Lux (and `main`) declares `SimpleChains = "0.4.7"`, which hard-blocks resolution in the docs env (`Unsatisfiable requirements ... restricted by compatibility requirements with Lux to versions: 0.4.7 - 0.4.8`). The one-line widen is verified safe: Lux `main` + SimpleChains v0.5.0 runs `ToSimpleChainsAdaptor` forward and Zygote gradient correctly.
2. Merge https://github.com/SciML/SciMLSensitivity.jl/pull/1611 (docs compat `SimpleChains = "0.4, 0.5"`).
3. Bump `docs/Manifest.toml` to SimpleChains 0.5 once a Lux release allows it — follow-up PR then closes this issue.

No change to the example text is needed — the fully Enzyme-based version from https://github.com/SciML/SciMLSensitivity.jl/pull/1502 works as written once SimpleChains 0.5 resolves.

**Comment by ChrisRackauckas-Claude** [comment]
Step 1 is now in motion: https://github.com/LuxDL/Lux.jl/pull/1748 widens Lux's SimpleChains compat to admit 0.5 (verified against Lux's own SimpleChains integration testset: 31 pass / 4 pre-marked broken / 0 unexpected failures with SimpleChains v0.5.0). Once that merges and a Lux release is tagged, the `docs/Manifest.toml` bump PR here follows.

**Comment by ChrisRackauckas-Claude** [comment]
## Status: the fix is done and verified; every remaining blocker is in Lux

**The original bug is fixed upstream.** SimpleChains 0.5.0 (released 2026-08-25) ships the EnzymeCore extension from PumasAI/SimpleChains.jl#220, which makes the chain the differentiation boundary and removes the `IllegalTypeAnalysisException`. Verified against SciMLSensitivity `master` with an isolated Documenter build of the actual `docs/src/examples/neural_ode/simplechains.md` (Julia 1.12.7, SimpleChains v0.5.0, Enzyme v0.13.199, `warnonly = false`): all 300 `AutoEnzyme` training steps ran, loss `225.3562f0` → `0.35378605f0`, HTML rendered. **The page needs no edits** — the fully Enzyme-based version from https://github.com/SciML/SciMLSensitivity.jl/pull/1502 works as written once SimpleChains 0.5 resolves.

### Nothing further is needed in this repo

- https://github.com/SciML/SciMLSensitivity.jl/pull/1611 (dependabot) widens the docs compat to `SimpleChains = "0.4, 0.5"` — correct and ready.
- https://github.com/SciML/SciMLSensitivity.jl/pull/1614 fixes the `typos` job that reds every PR here, including 1611.
- After a Lux release lands, one more PR bumps `docs/Manifest.toml` (currently pinning SimpleChains v0.4.8) and this issue closes.

### Why Lux is the whole remaining blocker

The docs environment contains Lux, and **every released Lux — and `main` — pins `SimpleChains = "0.4.7"`**, so SimpleChains 0.5 cannot resolve there at all:

```
ERROR: Unsatisfiable requirements detected for package SimpleChains [de6bee2f]:
 └─restricted by compatibility requirements with Lux [b2108857] to versions: 0.4.7 - 0.4.8 — no versions left
```

Two PRs are open against Lux:

1. **https://github.com/LuxDL/Lux.jl/pull/1748** — widen Lux's SimpleChains compat to `"0.4.7, 0.5"` (package, downstream-test project, example project). Verified with Lux's own SimpleChains integration testset on SimpleChains v0.5.0: **31 pass / 4 pre-marked broken / 0 unexpected failures**.
2. **https://github.com/LuxDL/Lux.jl/pull/1749** — repair Lux's CI, which is red on every PR for reasons unrelated to (1). The six `downgrade` jobs abort precompiling `ReactantCUDAExt` (`LLVM error: unknown function pass 'custom-nvvm-reflect'`); auditing every declared floor against every other floor's requirements found **twelve stale pins**, including three capping `ForwardDiff` at 0.10 while Lux floors it at 1, and two capping `AMDGPU` below the declared 2.2 floor. The corrected set resolves and precompiles with the local `lib/*` packages developed, the way CI builds the environment.

So the sequence is: Lux merges 1749 and 1748 → tags a release → dependabot/CompatHelper or a one-line PR bumps `docs/Manifest.toml` here → this issue closes. Nothing in the sequence is in SciML's hands, which is why I am leaving this open rather than closing it on the upstream fix.

### Two Lux-side caveats worth recording

- **Lux CI never exercises the SimpleChains path.** `test/runtests.jl` gates integration tests on `VERSION ≥ v"1.11-"` **and** group `others`, then skips SimpleChains when `VERSION ≥ v"1.12-"`; the matrix is 1.10 and 1.12, so the file runs on neither. Evidence for 1748 is therefore local, not CI.
- **A separate Enzyme/GPUCompiler break is live on Julia 1.12**, unrelated to SimpleChains: Enzyme 0.13.199 calls `GPUCompiler.runtime_slug` (`Enzyme/src/compiler.jl:181`), which no longer exists. It is what fails `test (1.12, others)` in Lux, and it also affects https://github.com/LuxDL/Lux.jl/pull/1739. There is no newer Enzyme to move to, so it needs an upper bound or an upstream fix rather than a floor bump.

**Comment by ChrisRackauckas-Claude** [comment]
## 2026-08-31 clean-master recheck

I reproduced this again in a fresh detached worktree at exact SciMLSensitivity `master` commit `d3c23ca6d5934f9c997036ad0cb2cfaf383da926`.

The focused docs example used:

```sh
timeout 3600 "$HOME/.juliaup/bin/julia" +1.12.7 --project=docs enzyme_repro.jl
```

with a newly resolved docs environment:

```text
SciMLSensitivity 7.119.2
Enzyme 0.13.199
SimpleChains 0.4.8
Optimization 5.9.0
OrdinaryDiffEq 7.8.1
Julia 1.12.7 (6d172b025e4)
Linux x86_64, LLVM 18.1.7
```

`maxiters = 1` fails before the first optimization step with:

```text
IllegalTypeAnalysisException: Enzyme compilation failed due to illegal type analysis.
Failure within method:
(::SimpleChains.SimpleChain{...})(::Vector{Float32}, ::Vector{Float32})

Caused by:
 [1] reinterpret          @ Base ./essentials.jl:733
 [2] align                @ VectorizationBase/src/alignment.jl:13
 [3] with_stack_memory    @ SimpleChains/src/memory.jl:25
 [4] SimpleChain          @ SimpleChains/src/simple_chain.jl:182
```

The failure still reduces to Enzyme + SimpleChains only; no SciML package is needed:

```julia
import Enzyme
import SimpleChains

sc = SimpleChains.SimpleChain(
    SimpleChains.static(2),
    SimpleChains.Activation(x -> x .^ 3),
    SimpleChains.TurboDense{true}(tanh, SimpleChains.static(50)),
    SimpleChains.TurboDense{true}(identity, SimpleChains.static(2)),
)
x = Float32[2.0, 0.0]
p = Array(SimpleChains.init_params(sc))
dp = zero(p)
loss(p) = sum(abs2, sc(x, p))

Enzyme.autodiff(
    Enzyme.set_runtime_activity(Enzyme.Reverse),
    loss,
    Enzyme.Active,
    Enzyme.Duplicated(p, dp),
)
```

On Julia 1.12.7 with Enzyme 0.13.199 and SimpleChains 0.4.8, that fails with the same `reinterpret -> align -> with_stack_memory` illegal type analysis. In a fresh two-dependency environment with **SimpleChains 0.5.0**, the exact same script exits 0:

```text
Enzyme 0.13.199
SimpleChains 0.5.0
Julia Version 1.12.7
SUCCESS gradient_norm=34.95609
```

This confirms the fixed boundary is SimpleChains 0.5.0, which contains the reviewed custom Enzyme rule from https://github.com/PumasAI/SimpleChains.jl/pull/220. The original upstream diagnosis remains https://github.com/EnzymeAD/Enzyme.jl/issues/3292.

### Remaining resolver blocker

I tested the prospective SciMLSensitivity change `SimpleChains = "0.5"`. A fresh docs instantiation fails before tests:

```text
ERROR: Unsatisfiable requirements detected for package SimpleChains [de6bee2f]:
 ├─restricted to versions 0.5 by project, leaving only 0.5.0
 └─restricted by compatibility requirements with Lux to versions 0.4.7 - 0.4.8 or uninstalled — no versions left
```

Lux's compatibility fix has now merged at https://github.com/LuxDL/Lux.jl/pull/1748 (merge commit `83f42c78a435f59566f9af6e64f8ccfddafbb6ca`), but as of 2026-08-31 the latest registered Lux release/tag is still `v1.31.4`, which predates that merge. Consequently there is no satisfiable, released docs environment that combines Lux with SimpleChains 0.5 yet.

The follow-up after Lux publishes a release containing that merge is:

1. Require SimpleChains 0.5 in `docs/Project.toml` so the known-bad 0.4.8 implementation cannot resolve.
2. Instantiate the docs environment and confirm it selects the new Lux release plus SimpleChains 0.5.0.
3. Run the focused SimpleChains neural-ODE example and the full docs build, then close this issue.

I am not proposing `Enzyme.API.strictAliasing!(false)`: Enzyme documents `API` as internal, and the Enzyme maintainer confirmed the type-punned SimpleChains stack allocation requires a custom package rule rather than disabling analysis. No SciMLSensitivity change is safe until the Lux release exists.
